oeps/oep-0012.rst

```diff
@@ -0,0 +1,520 @@
+but this requires that the feature add a new URL for the feature, and render the
+whole page at that URL using a Mako template. It would be preferable for a
+course tab to be able to register a pluggable user interface that will be
+presented as a tab, and for the implementor not to need to know how to implement
+the full page.
+
+There has been a long standing request to support rendering XBlocks as course
+tabs (`Allow xblocks to be shown as course tabs`_), but there are conceptual
+issues with this approach:
+
+* XBlocks must be configured and saved into the modulestore which doesn't make
+  much sense for course tabs that don't have their own state
+
+* Replacing a built-in tab such as "Course" would require a data migration for
+  every course if it was saved as an XBlock
+
+* It seems confusing to make an XBlock that should never be added into the
+  courseware, but which only exists to be shown as a tab
+
+It would be prefable to allow a feature library to register a course tab plugin
+which in turn renders using a component. This allows the feature to provide a
+custom UI without having to add custom blocks to each course, and if the feature
+is removed then the tab will seamlessly disappear without having to remove the
+blocks again.
+
+Use Case 2: LMS Discussions
+---------------------------
+
+Open edX Discussions is a complex feature that integrates into many parts of the
+LMS (currently as an inline XBlock, as a courseware tab, and to provide team
+discussions). Ideally, Open edx Discussions would be a separately-installable
+feature that would provide plugins for all of this functionality, and then an
+Open edX operator could choose whether to install edX discussions or an
+alternate collaboration tool.
+
+Currently the discussion board is a Django view built into edx-platform, but it
+should instead be made into a plugin. One option would be to create a new
+CourseDiscussionXBlock to display the "Course" tab, but as described above there
+are downsides to forcing this to be an XBlock. It would be cleaner if the edX
+discussion library provided a pluggable user interface that can be dropped into
+each of the contexts in which it is needed. The component would be provided
+with the context for authorization and to customize itself appropriately.
+
+An interesting question is how such components can share course-scoped settings
+that the instructor specified in Studio. See `OEP-6 Context-scoped XBlock
+Fields`_ for the design of context-scoped fields in XBlock.
+
+Use Case 3: Insights Extensions
+-------------------------------
+
+EdX Insights is a course analytics application that could be extended in
+a variety of different ways. Some examples include:
+
+1. ** New Charts **
+
+   A plugin could provide new types of charts for existing data types, or it
+   could extend the application to show charts for data that had not previously
+   been visualized.
```

Comment:
This was actually a use case in the design of XBlocks. The idea was that Insights would act as a data service, and XBlocks would grab data and render it. We could then define a front end as a set of XBlocks. Adding component would force us to reinvent solutions to many problems -- for example, we'd need some mechanism for determining which views to render in the dashboard. We would also want some way to compose dashboards from such renderings (we may want more than one per page). OLX lets us do that, as do nested XBlocks. Here, it's the same problem, just pushed into the future.